File: src/main/java/com/morak/performancetracker/aop/util/ProxyFactoryBeanUtils.java
package com.morak.performancetracker.aop.util;

import java.lang.reflect.Proxy;
import org.springframework.aop.PointcutAdvisor;
import org.springframework.aop.framework.ProxyFactoryBean;

public class ProxyFactoryBeanUtils {

    private ProxyFactoryBeanUtils() {
    }

    public static Object createObject(Object target, boolean setProxyTargetClass, PointcutAdvisor advisor) {
        ProxyFactoryBean proxyFactoryBean = new ProxyFactoryBean();
        proxyFactoryBean.setTarget(target);
        proxyFactoryBean.setProxyTargetClass(setProxyTargetClass);
        proxyFactoryBean.addAdvisor(advisor);
        return proxyFactoryBean.getObject();
    }
}

File: src/main/kotlin/com/morak/performancetracker/aop/util/ProxyFactoryBeanUtils.kt
package com.morak.performancetracker.aop.util

import org.springframework.aop.PointcutAdvisor
import org.springframework.aop.framework.ProxyFactoryBean

object ProxyFactoryBeanUtils {
    fun createObject(target: Any, setProxyTargetClass: Boolean, advisor: PointcutAdvisor): Any {
        val proxyFactoryBean = ProxyFactoryBean()
        proxyFactoryBean.setTarget(target)
        proxyFactoryBean.isProxyTargetClass = setProxyTargetClass
        proxyFactoryBean.addAdvisor(advisor)
        return proxyFactoryBean.getObject() ?: throw IllegalArgumentException("Fails to create proxy")
    }
}
